generator/template.go

```diff
@@ -195,6 +196,129 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
     }
   })
 }
+
+type Primitive = string | boolean | number;
+type RequestPayload = Record<string, unknown>;
+type FlattenedRequestPayload = Record<string, Primitive | Array<Primitive>>;
+
+/**
+ * Checks if given value is a plain object
+ * source: https://github.com/char0n/ramda-adjunct/blob/master/src/isPlainObj.js
+ * @param  {unknown} value
+ * @return {boolean}
+ */
+function isPlainObject(value: unknown): boolean {
+  const isObject =
+    Object.prototype.toString.call(value).slice(8, -1) === "Object";
+  const isObjLike = value !== null && isObject;
+
+  if (!isObjLike || !isObject) {
+    return false;
+  }
+
+  const proto = Object.getPrototypeOf(value);
+
+  const hasObjectConstructor =
+    typeof proto === "object" &&
+    proto.constructor === Object.prototype.constructor;
+
+  return hasObjectConstructor;
+}
+
+/**
+ * Checks if given value is of a primitive type
+ * @param  {unknown} value
+ * @return {boolean}
+ */
+function isPrimitive(value: unknown): boolean {
+  return ["string", "number", "boolean"].some(t => typeof value === t);
+}
+
+/**
+ * Checks if given primitive is zero-value
+ * @param  {Primitive} value
+ * @return {boolean}
+ */
+function isZeroValuePrimitive(value: Primitive): boolean {
+  return ["", "0", "false"].some(zv => value.toString() === zv);
+}
+
+/**
+ * Flattens a deeply nested request payload and returns an object
+ * with only primitive values and non-empty array of primitive values
+ * as per https://github.com/googleapis/googleapis/blob/master/google/api/http.proto
+ * @param  {RequestPayload} requestPayload
+ * @param  {String} path
+ * @return {FlattenedRequestPayload>}
+ */
+function flattenRequestPayload<T extends RequestPayload>(
+  requestPayload: T,
+  path: string = ""
+): FlattenedRequestPayload {
+  return Object.keys(requestPayload).reduce(
+    (acc: T, key: string): T => {
+      const value = requestPayload[key];
+      const newPath = path ? [path, key].join(".") : key;
+
+      const isNonEmptyPrimitiveArray =
+        Array.isArray(value) &&
+        value.every(v => isPrimitive(v)) &&
+        value.length > 0;
+
+      let objectToMerge = {};
+
+      if (isPlainObject(value)) {
+        objectToMerge = flattenRequestPayload(value as RequestPayload, newPath);
+      } else if (
+        isPrimitive(value) &&
+        !isZeroValuePrimitive(value as Primitive)
+      ) {
+        objectToMerge = { [newPath]: value };
+      } else if (isNonEmptyPrimitiveArray) {
+        const nonEmptyPrimitiveArray = value as Primitive[];
+        objectToMerge = {
+          [newPath]: nonEmptyPrimitiveArray.filter(
+            m => !isZeroValuePrimitive(m)
+          )
+        };
+      }
```

Comment:
I am filtering zero value primitives here

Reply:
was gonna say, empty values in the array is the case that you shouldn't filter.

Reply:
Ah I see hmmm... but if we don't filter them then they will get added to the path as URL query params? Is that OK?

Reply:
there's no default value in the repeated array. the default value case here only applies for primitive fields. array no need to touch.

e.g. `a: [1,0,1]` will turn  into `a=1&a=0&a=1`,  if you put `a=1&a=1` then in the backend you only have `a:[1,1]`

Reply:
Ah I see. I understand what you mean. Will fix this right away 🙇🏼

Reply:
> it's not needed though. and jsonpb also ignores the default value to save space. we should follow the same.
> https://developers.google.com/protocol-buffers/docs/proto3#json

I was giving this some consideration. The above documentation also says this
```
 An implementation may provide options to emit fields with default values in the JSON-encoded output.
```
Maybe we can add a flag like golang's grpc gateway plugin `EmitDefault` and if it is set to true, we filter zero value primitives otherwise we don't filter them. Let me know what you think about this 🙏🏼 

Reason is as follows:
Lets take the following payload where `b` & `c` are part of the query parameters `/api/query/a?msg.b=false&c=hello`
```
{
  "a": 10
  "msg": { "b": false },
  "c": hello
}
```

The user may intentionally want to save `false` in the database but if we remove it from query params `/api/query/a?c=hello` and try to read the `msg` object on the server side, it will be of value `nil` and we can't be really sure whether the user wanted to save `false` or `null` to the database.

Reply:
This is the option in `grpc-gateway` I am talking about. `EmitUnpopulated`. We can perhaps add a similar option and let the user decide whether they want to filter zero-values or not.

https://grpc-ecosystem.github.io/grpc-gateway/docs/development/grpc-gateway_v2_migration_guide/#we-now-emit-default-values-for-all-fields

Reply:
grpc-gateway's emitting the default 

https://github.com/grpc-ecosystem/grpc-gateway/blob/master/runtime/marshaler_registry.go#L23

Reply:
Yes, as of version 2 they are emitting the default values but as the user of the package we can decide whether we want to enable it to true or false.
Similarly we can have an option here like `EmitUnpopulated` where we can filter default-value fields if set to false and not filter them if set to true.

Reply:
I have revised the code as per your suggestions. Maybe `EmitUnpopulated` can be a different PR?

Reply:
had a second read. what we are talking about here is unmarshalling because the plugin sends a json and grpc gateway uses `protojson` to unmarshal the payload into the generated proto struct. 

Then we were looking at the wrong place. `EmitUnpopulated` affects payload on the way down, not on the way up to grpc-gateway.

So we will always have the default value as per described here.
https://developers.google.com/protocol-buffers/docs/proto3#default

like I mentioned before, it's a thing that's worth a sentence/short paragraph to describe the behaviour.

Reply:
Yes, I did a couple of reads and also the above link and came to the same conclusion.
My apologies for the misunderstanding and the long discussion regarding this 🙇🏼 
Should we describe this behaviour in the README or as a comment?

Reply:
yes please. a section call `Notes` before the `Example` will be good. just number dotted the title and have a few sentences about it.

Reply:
Understood. Will do so and update this PR and thank you once again for the patience in reviewing my code and giving me this chance to making this contribution 🙇🏼